### PR TITLE
gazebo_plugins Logical camera plugin

### PR DESCRIPTION
**Pull request by irsdkv**
Hi all
I didnt found `LogicalCameraImage` message on ROS2 package so I placed in to `gazebo_msgs/msg/LogicalCameraImage.msg`
I haven't enough knowledge how I shall to implement this whole functionality in ROS2 way so I will be grateful if someone let me know how it shall be done.

### CODE REVIEW

File: gazebo_plugins/CMakeLists.txt [hunk 1]
```diff
@@ -400,6 +411,7 @@ install(TARGETS
     gazebo_ros_bumper
     gazebo_ros_camera
     gazebo_ros_depth_camera
+    gazebo_ros_logical_camera
```
Comment (chapulina) [inline gazebo_plugins/CMakeLists.txt]: Nit: keep alphabetical order

File: gazebo_plugins/CMakeLists.txt [hunk 2]
```diff
@@ -376,6 +376,17 @@ ament_target_dependencies(gazebo_ros_projector
 )
 ament_export_libraries(gazebo_ros_projector)
 
+add_library(gazebo_ros_logical_camera SHARED
+  src/gazebo_ros_logical_camera.cpp
+)
+ament_target_dependencies(gazebo_ros_logical_camera
+  "gazebo_ros"
+  "sensor_msgs"
```
Comment (chapulina) [inline gazebo_plugins/CMakeLists.txt]: Is `sensor_msgs` needed?

File: gazebo_plugins/include/gazebo_plugins/gazebo_ros_logical_camera.hpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
```
Comment (chapulina) [inline gazebo_plugins/include/gazebo_plugins/gazebo_ros_logical_camera.hpp]:
```suggestion
// Copyright 2020 Open Source Robotics Foundation, Inc.
```

File: gazebo_plugins/src/gazebo_ros_logical_camera.cpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  auto topic_name = this->impl_->sensor_->Topic();
+
+  impl_->pub_ =
+    impl_->ros_node_->create_publisher<gazebo_msgs::msg::LogicalCameraImage>(
+    topic_name, rclcpp::SensorDataQoS());
+
+  RCLCPP_INFO(impl_->ros_node_->get_logger(), "Created [%s] topic", impl_->pub_->get_topic_name());
+
+  // Create message to be reused
+  auto msg = std::make_shared<gazebo_msgs::msg::LogicalCameraImage>();
```
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_logical_camera.cpp]: I think you can prevent a copy by removing the temporary `msg` object, right?

**Review by chapulina — COMMENTED** [review]
Thank you for the contribution! It's great to have this ROS 2 interface to the logical camera. I just have some minor comments below.

**Comment by irsdkv** [comment]
Thank you!
All done

**Comment by irsdkv** [comment]
Hi!
Any update?
I want to create PR to PX4 [uRTPS repo ](https://github.com/PX4/micrortps_agent) with bugfix which related to MoCap and functionality of this PR

**Review by chapulina — APPROVED** [review]
Thanks for the contribution, this works for me! You should think of announcing the new plugin on https://community.gazebosim.org/ :wink:

**Pull request merged**